File: 538.把二叉搜索树转换为累加树.cpp
/*
 * @lc app=leetcode.cn id=538 lang=cpp
 *
 * [538] 把二叉搜索树转换为累加树
 */

#include <algorithm>
#include <iostream>
#include <vector>
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
private:
  void dfs(std::vector<TreeNode *> &nodes, TreeNode *root) {
    if (root) {
      nodes.push_back(root);
      dfs(nodes, root->left);
      dfs(nodes, root->right);
    }
  }

public:
  /**
   * @brief
   * 215/215 cases passed (36 ms)
   * Your runtime beats 57.83 % of cpp submissions
   * Your memory usage beats 5.02 % of cpp submissions (33.4 MB)
   * @param root
   * @return TreeNode*
   */
  TreeNode *convertBST(TreeNode *root) {
    std::vector<TreeNode *> nodes;
    dfs(nodes, root);
    std::sort(nodes.begin(), nodes.end(),
              [](TreeNode *a, TreeNode *b) { return a->val < b->val; });
    for (int i = nodes.size() - 2; i >= 0; i--) {
      nodes[i]->val += nodes[i + 1]->val;
    }
    return root;
  }
};
// @lc code=end

File: 538.把二叉搜索树转换为累加树.java
/*
 * @lc app=leetcode.cn id=538 lang=java
 *
 * [538] 把二叉搜索树转换为累加树
 */
import java.util.ArrayList;
import java.util.Collections;

class TreeNode {

  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {}

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

  private ArrayList<TreeNode> dfs(ArrayList<TreeNode> nodes, TreeNode root) {
    if (root != null) {
      nodes.add(root);
      return dfs(dfs(nodes, root.left), root.right);
    } else {
      return nodes;
    }
  }

  /**
   * 215/215 cases passed (4 ms)
   * Your runtime beats 7.35 % of java submissions
   * Your memory usage beats 80.81 % of java submissions (41.4 MB)
   * @param root
   * @return
   */
  public TreeNode convertBST(TreeNode root) {
    var nodes = new ArrayList<TreeNode>();
    nodes = dfs(nodes, root);
    Collections.sort(nodes, (a, b) -> a.val - b.val);
    for (int i = nodes.size() - 2; i >= 0; i--) {
      nodes.get(i).val += nodes.get(i + 1).val;
    }
    return root;
  }
}
// @lc code=end
